File: PA25.java
// LeetCode 238. Product of Array Except Self
//Time Complexity: O(n) //O(n+n) n is the no of elements in nums
//Space Complexity : O(1) // As output array doest count 
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] res= new int[nums.length];
        
        res[0] = 1;
        
        for(int i=1; i<nums.length ; i++){
            res[i]=nums[i-1]*res[i-1];  
        }
        int temp=1;
        for(int k=nums.length-1; k>=0 ;k--){
            res[k]= temp*res[k];
            temp = temp*nums[k];
        }
        return res;
    }
}

File: PA_25.py
#LC 238 - Product of Array Except Self
#Time Complexity - O(2n)
#Space Complexity - O(1)
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        if not nums:
            return nums
        tmp = [1] * len(nums)
        for i in range(1,len(nums)):
            tmp[i] = tmp[i-1]*nums[i-1]
        
        i = len(nums)-2
        r = nums[i+1]
        while i >= 0:
            tmp[i] = tmp[i] * r
            r = r * nums[i]
            i = i - 1
        return tmp